### PR TITLE
Fix/issue 38 password keyboard behaviour

### PR DESCRIPTION
**Pull request by Basler182**
# *Fix/issue 38 password keyboard behaviour*

## :recycle: Current situation & Problem
#38 

## :gear: Release Notes 

- Added Password Toogle Icon to Login Screen
- Added Login Form Validator
- Added Login Error Cases 
- Switched to OutlinedTextFields

## :white_check_mark: Testing
- Added CredentialRegisterManagerAuthTest
- Added LoginViewModelTest
- Added LoginFormValidatorTest

## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by PSchmiedmayer** [comment]
Thank you for taking a look at this and fixing the issues in smaller targeted PRs @Basler182 🚀
I will defer to @eldcn or @pauljohanneskraft for an in-depth review of the changes.

File: core/design/src/main/kotlin/edu/stanford/spezi/core/design/component/validated/outlinedtextfield/ValidatedOutlinedTextField.kt [hunk 1]
```diff
@@ -31,17 +37,23 @@ fun ValidatedOutlinedTextField(
...
-            singleLine = true,
-            keyboardOptions = KeyboardOptions.Default.copy(imeAction = ImeAction.Next),
+            singleLine = singleLine,
+            keyboardOptions = keyboardOptions,
             isError = errorText != null,
+            readOnly = readOnly,
+            trailingIcon = trailingIcon,
+            keyboardActions = keyboardActions,
+            visualTransformation = visualTransformation,
+            supportingText = if (errorText != null) {
```
Comment (eldcn) [inline core/design/src/main/kotlin/edu/stanford/spezi/core/design/component/validated/outlinedtextfield/ValidatedOutlinedTextField.kt:47]: Minor: `supportingText = errorText?.let { Text(text = it) },`

File: modules/account/src/main/kotlin/edu/stanford/spezi/module/account/register/RegisterViewModel.kt [hunk 2]
```diff
@@ -43,16 +45,26 @@ class RegisterViewModel @Inject internal constructor(
                 }
 
                 is Action.DateFieldUpdate -> {
-                    val updatedUiState = it.copy(dateOfBirth = action.newValue)
+                    val deviceLocale: Locale = Locale.getDefault()
+                    val updatedUiState = it.copy(
+                        dateOfBirth = action.newValue,
+                        formattedDateOfBirth = action.newValue.format(
+                            DateTimeFormatter.ofPattern(
```
[the remark this answers is not in the document]
Comment (Basler182) [reply modules/account/src/main/kotlin/edu/stanford/spezi/module/account/register/RegisterViewModel.kt:52]: good idea! 💯

File: modules/account/src/main/kotlin/edu/stanford/spezi/module/account/login/LoginFormValidator.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    } else {
+        Result.Invalid("Password must be at least $MIN_PASSWORD_LENGTH characters")
+    }
+
+    fun isFormValid(uiState: UiState): Boolean {
+        return emailResult(uiState.email.value) is Result.Valid &&
+            passwordResult(uiState.password.value) is Result.Valid
+    }
+
+    fun isEmailValid(email: String): Boolean = emailResult(email) is Result.Valid
```
[the remark this answers is not in the document]
Comment (Basler182) [reply modules/account/src/main/kotlin/edu/stanford/spezi/module/account/login/LoginFormValidator.kt:25]: Agreed with that.

File: modules/account/src/main/kotlin/edu/stanford/spezi/module/account/login/LoginViewModel.kt [hunk 4]
```diff
@@ -40,33 +43,57 @@ class LoginViewModel @Inject constructor(
                     it
                 }
 
-                is Action.GoogleSignIn -> {
-                    googleSignIn()
+                is Action.GoogleSignInOrSignUp -> {
```
[the remark this answers is not in the document]
Comment (Basler182) [reply modules/account/src/main/kotlin/edu/stanford/spezi/module/account/login/LoginViewModel.kt:46]: Good improvement. 🚀

File: modules/account/src/main/kotlin/edu/stanford/spezi/module/account/login/LoginViewModel.kt [hunk 5]
```diff
@@ -40,33 +43,57 @@ class LoginViewModel @Inject constructor(
...
+        }
+
+        if (!uiState.isAlreadyRegistered && validator.isFormValid(uiState)) {
+            navigateToRegister()
+        }
+
+        _uiState.update {
+            it.copy(
+                hasAttemptedSubmit = true,
+                email = FieldState(
```
[the remark this answers is not in the document]
Comment (Basler182) [reply modules/account/src/main/kotlin/edu/stanford/spezi/module/account/login/LoginViewModel.kt:85]: @eldcn yes, cause we always just update the state value, but in this case we also want to update the state error. but i can use the new learned `it.email.copy(error=..)`

File: modules/account/src/main/kotlin/edu/stanford/spezi/module/account/login/UiState.kt [hunk 6]
```diff
@@ -1,12 +1,16 @@
...
-    val password: String = "",
-    val email: String = "",
+    val password: FieldState = FieldState(),
+    val email: FieldState = FieldState(),
     val passwordVisibility: Boolean = false,
     val showProgress: Boolean = false,
     val showFilterByAuthorizedAccounts: Boolean = true,
+    val isFormValid: Boolean = false,
     val isAlreadyRegistered: Boolean = false,
+    val hasAttemptedSubmit: Boolean = false,
```
[the remark this answers is not in the document]
Comment (Basler182) [reply modules/account/src/main/kotlin/edu/stanford/spezi/module/account/login/UiState.kt:13]: Good idea!

File: modules/account/src/main/kotlin/edu/stanford/spezi/module/account/login/LoginScreen.kt [hunk 7]
```diff
@@ -111,14 +162,10 @@ You may login to your existing account or create a new one if you don't have one
...
-                if (uiState.isAlreadyRegistered) {
-                    onAction(Action.PasswordCredentialSignIn)
-                } else {
-                    onAction(Action.NavigateToRegister)
-                }
+                onAction(Action.PasswordSignInOrSignUp)
             },
             modifier = Modifier.fillMaxWidth(),
-            enabled = uiState.email.isNotEmpty() && uiState.password.isNotEmpty()
+            enabled = uiState.email.value.isNotEmpty() && uiState.password.value.isNotEmpty()
```
[the remark this answers is not in the document]
Comment (Basler182) [reply modules/account/src/main/kotlin/edu/stanford/spezi/module/account/login/LoginScreen.kt:168]: Great improvement to the function.

File: modules/account/src/test/java/edu/stanford/spezi/module/account/login/LoginFormValidatorTest.kt [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    @Test
+    fun `given valid email when emailResult is called then return Valid`() {
+        // Given
+        val validEmail = "test@test.com"
+
+        // When
+        val result = loginFormValidator.emailResult(validEmail)
+
+        // Then
+        assertThat(result).isEqualTo(FormValidator.Result.Valid)
```
[the remark this answers is not in the document]
Comment (Basler182) [reply modules/account/src/test/java/edu/stanford/spezi/module/account/login/LoginFormValidatorTest.kt:21]: Great improvement, very satisfying. I didnt like the `        assertThat(result.isValid).isInstanceOf(FormValidator.Result.Invalid::class.java)` assertions

File: modules/account/src/main/kotlin/edu/stanford/spezi/module/account/login/LoginViewModel.kt [hunk 9]
```diff
@@ -85,37 +113,52 @@ class LoginViewModel @Inject constructor(
...
+                isFormValid = validator.isFormValid(uiState),
+                isPasswordSignInEnabled = newValue.value.isNotEmpty() && uiState.password.value.isNotEmpty()
+            )
         }
     }
 
     private fun forgotPassword() {
-        if (uiState.value.email.isEmpty()) {
-            messageNotifier.notify("Please enter your email")
+        if (validator.isValidEmail(uiState.value.email.value) is FormValidator.Result.Valid) {
```
Comment (eldcn) [inline modules/account/src/main/kotlin/edu/stanford/spezi/module/account/login/LoginViewModel.kt:149]: Can we update all usages of `is Result.Valid/Invalid` with `.isValid/.isValid.not()` as well? A simple string would show them all :)

File: modules/account/src/main/kotlin/edu/stanford/spezi/module/account/register/FormValidator.kt [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        const val MIN_PASSWORD_LENGTH = 6 // Minimum for firebase
+    }
+
+    sealed interface Result {
+        data object Valid : Result
+        data class Invalid(val message: String) : Result
+
+        fun errorMessageOrNull() = if (this is Invalid) message else null
+
+        val isValid: Boolean
```
Comment (eldcn) [inline modules/account/src/main/kotlin/edu/stanford/spezi/module/account/register/FormValidator.kt:33]: Tiny 😄 Let's in general put properties before methods

File: modules/account/src/main/kotlin/edu/stanford/spezi/module/account/register/RegisterFormValidator.kt [hunk 11]
```diff
@@ -36,27 +33,24 @@ class RegisterFormValidator @Inject internal constructor() {
...
-    sealed interface Result {
-        data object Valid : Result
-        data class Invalid(val message: String) : Result
-
-        fun errorMessageOrNull() = if (this is Invalid) message else null
-    }
-
-    private companion object {
-        const val MIN_PASSWORD_LENGTH = 6 // Minimum for firebase
+    fun isRegisterButtonEnabled(uiState: RegisterUiState): Boolean {
```
Comment (eldcn) [inline modules/account/src/main/kotlin/edu/stanford/spezi/module/account/register/RegisterFormValidator.kt:49]: This method should belong to the view model, validator is responsible to validate the form and not to decide whether the button should be enabled or not ;)

**Review by eldcn — APPROVED** [review]
🚀

**Pull request merged**